proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,208 @@
+          runtime_type = "io.containerd.spin-v1.2.3.v1"
+        [plugins.cri.containerd.runtimes.spin-v1_2_3]
+          runtime_type = "io.containerd.spin-v1.2.3.v1"
+      ```
+3. A runtime class that contains a handler that matches the "spin" config runtime name from step 2.
+      ```yaml
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1.2.3
+      handler: spin-v1_2_3
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1.2.3: "true"
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1.2
+      handler: spin-v1_2
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1.2: "true"
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1
+      handler: spin-v1
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1: "true"
+      ```
+
+    **NOTE: node labels would be expected for both `wasmtime-spin-v1: "true"` and `wasmtime-spin-v1.2.3: "true"` on nodes that have the spin shim v1.2.3 installed.**
+
+4. A Wasm workload with the specified RuntimeClass name matching the "wasmtime-spin-v1" runtime class from step 3. The "wasmtime-spin-v1" runtimeClassName would map to the spin-v1.2.3 handler.
+      ```yaml
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: wasm-spin
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            app: wasm-spin
+        template:
+          metadata:
+            labels:
+              app: wasm-spin
+          spec:
+            runtimeClassName: wasmtime-spin-v1
+            containers:
+            - name: spin-hello
+              image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
+              command: ["/"]
+      ```
+
+### <a name='Introducinganewversionofashim'></a>Introducing a new version of a shim
+Now that we have described the versioning scheme, let's describe what happens when we introduce a new node that contains a new version of a shim. In this example, we will introduce v1.2.4 of the Spin shim.
+
+For this example, let's imagine we have a cluster with 1 node configured as described in [Versioning of both shim binary and RuntimeClass name](#versioning-of-both-shim-binary-and-runtimeclass-name). We introduce a new node to the cluster with `/bin/containerd-shim-spin-v1.2.4-v1` installed and the following containerd config section.
+
+```toml
+[plugins.cri.containerd.runtimes.spin-v1]
+runtime_type = "io.containerd.spin-v1.2.4.v1"
+[plugins.cri.containerd.runtimes.spin-v1_2]
+runtime_type = "io.containerd.spin-v1.2.4.v1"
+[plugins.cri.containerd.runtimes.spin-v1_2_4]
+runtime_type = "io.containerd.spin-v1.2.4.v1"
+```
+
+This new node running v1.2.4 would be decorated with the following node labels, `wasmtime-spin-v1`, `wasmtime-spin-v1.2`, and `wasmtime-spin-v1.2.4`. 
```

Comment:
I think this _probably_ makes sense in the general case, but I suspect you can be a lot more targeted when talking about best practices. 

In the absolute worst case since we are building on top of unstable APIs, let's say a patch version does have breaking changes. A service offering these runtimes would need to offer both patch versions side by side during migration. 

other than that, I would imagine exposing all the way down to the patch version is not necessarily useful duplication (e.g. not sure I follow the motivation for v1, v1.2, and v1.2.4).

and if we did have a patch version that broke things, how are these aliases supposed to work? e.g. I have `v1.2.3 + v1.2.4`, should `v1` **change** from `v1.2.3` to `v1.2.4`? That seems reasonable but a bit overly clever and would result in breaking users who opted to specify the major/minor version only. 

I think the naming probably makes sense, but without aliasing. The runtime class should expose the most precise naming possible, while the node labels _may_ still need the aliasing for scheduling/compat during migration

Reply:
I also just realized I only saw the doc comments, not the PR comments :) might be some redundancy with existing comments

Reply:
> If there are features in v1.1 that an app is using, which would not allow it to run on v1.0, then we need to inform the scheduler that the workload is unable to run on v1.0 nodes and only on v1.1+. I believe we can do that effectively with node labels. The one caveat being that I don't know how to model a minimum version constraint; the scenario where an app that needs min shim version v1.1 and could be schedule on any node supporting v1.1+. For example, I don't see a way to tell the scheduler select a node that has a label greater than v1.1.

yeah so this is what I mean w.r.t the runtime classes and node labels. I think this is where the "aliasing" (or, not really aliasing, but capability discovery) would still be useful.

Node has labels:
```
wasmtime-spin-enabled: "true"
wasmtime-spin-v1.2: "true"
```

Add support for new runtime class
```
wasmtime-spin-enabled: "true"
wasmtime-spin-v1.2: "true"
wasmtime-spin-v1.3: "true"
```

Remove original runtime class
```
wasmtime-spin-enabled: "true"
wasmtime-spin-v1.3: "true"
```

If you didn't actually care about features in 1.2 vs 1.3, you could select on `wasmtime-spin-enabled` and let node rotation remove eligible nodes for `1.2` naturally. If you cared about versions, you could select them explicitly.

I tend to think major/minor is enough granularity, but could see an argument for patch given the stability.

Reply:
I've updated based on your feedback. Thank you!